src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/OpenDocumentGenerator.cs

```diff
@@ -0,0 +1,317 @@
+                    }
+                    catch (Exception ex)
+                    {
+                        ReportError(ex);
+                    }
+                }
+
+                OnCompletingBackgroundWork();
+
+                await Task.Factory.StartNew(
+                    () =>
+                    {
+                        NotifyDocumentsProcessed(work);
+                    },
+                    CancellationToken.None,
+                    TaskCreationOptions.None,
+                    _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+                lock (_work)
+                {
+                    // Resetting the timer allows another batch of work to start.
+                    _timer.Dispose();
+                    _timer = null;
+
+                    // If more work came in while we were running start the worker again.
+                    if (_work.Count > 0)
+                    {
+                        StartWorker();
+                    }
+                }
+
+                OnCompletedBackgroundWork();
+            }
+            catch (Exception ex)
+            {
+                // This is something totally unexpected, let's just send it over to the workspace.
+                ReportError(ex);
+
+                _timer?.Dispose();
+                _timer = null;
+            }
+        }
+
+        private void NotifyDocumentsProcessed(KeyValuePair<string, DocumentSnapshot>[] work)
+        {
+            for (var i = 0; i < work.Length; i++)
+            {
+                foreach (var documentProcessedTrigger in _documentProcessedListeners)
+                {
+                    var document = work[i].Value;
+                    documentProcessedTrigger.DocumentProcessed(document);
+                }
+            }
+        }
+
+        private void ProjectSnapshotManager_Changed(object sender, ProjectChangeEventArgs args)
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            switch (args.Kind)
+            {
+                case ProjectChangeKind.ProjectChanged:
+                    {
+                        var projectSnapshot = args.Newer;
+                        foreach (var documentFilePath in projectSnapshot.DocumentFilePaths)
+                        {
+                            var document = projectSnapshot.GetDocument(documentFilePath);
+                            Enqueue(document);
+                        }
+
+                        break;
+                    }
+
+                case ProjectChangeKind.DocumentAdded:
+                    {
+                        var projectSnapshot = args.Newer;
+                        var document = projectSnapshot.GetDocument(args.DocumentFilePath);
+
+                        // We don't enqueue the current document because added documents are by default closed.
+
+                        foreach (var relatedDocument in projectSnapshot.GetRelatedDocuments(document))
```

Comment:
I wonder if it buys us anything to create a `GetRelatedDocuments` overload that takes just the TargetPath for the document to save us the Document lookup above.

Reply:
Hmmm we could, however the implication of it needing the "targetPath" and not the filepath would add that complexity to the caller. Given document creation isn't typically that expensive I'd like to have the overload as is for now.